## KL ダイバージェンス

### モチベーション

- 2 つの確率分布 $p(x),q(x)$ の間に，距離のようなものを定義したい
- 以下の条件を満たしてほしい

> **条件**
>
> 1. $p(x), q(x)$ が異なるほど大きい値をとる
> 2. $p(x) = q(x)$ のときのみ $0$ ，それ以外は正の値をとる

これを満たす関数が KL ダイバージェンスである．

### KL ダイバージェンスとは

KL ダイバージェンスとは，2 つの確率密度を測る尺度である．

2 つの確率分布 $p(x)$ と $q(x)$ の間の KL ダイバージェンスは以下のように定義される．

$$
D_\mathrm{KL}(p || q) = \int p(x) \log \frac{p(x)}{q(x)} dx
$$

### KL ダイバージェンスの特徴

- KL ダイバージェンスは上の条件 1,2 を満たす
- $D_\mathrm{KL}(p||q)$ と $D_\mathrm{KL}(q||p)$ は同じ値をとるとは限らない\
  → [距離の公理](https://ja.wikipedia.org/wiki/%E8%B7%9D%E9%9B%A2%E7%A9%BA%E9%96%93)を満たさないので，厳密には距離ではない

### 実際に検証してみる

In [2]:
// 依存関係のインストール
:dep image = "0.23"
:dep evcxr_image = "1.1"

// プロット用ライブラリ
:dep plotters = { version = "^0.3.5", default_features = false, features = ["evcxr", "all_series", "all_elements", "bitmap_backend", "full_palette", "point_series", "colormaps"] }

// 乱数
:dep rand = "0.8.5"
:dep rand_distr = "0.4.3"

// 配列
:dep ndarray = { version = "0.15.6" }

In [3]:
// インポート
use evcxr_image::ImageDisplay;
use image::{GenericImage, imageops::FilterType};

use plotters::prelude::*;
use rand::prelude::*;

use ndarray::{Array, ArrayView, Axis, array, s};

In [4]:
:dep myml = { path = "../myml/" }

In [5]:
use myml::utility::linspace;

#### 比較する確率分布

$$
\begin{align*}
    f(x) &= \frac{1}{\sqrt{2\pi}} \exp \left( -\frac{x^2}{2} \right) \quad (= \mathcal{N}(0, 1))\\[10pt]
    g(x) &=
    \left\{
    \begin{array}{ll}
        1/4 x + 1/2 & \mathrm{if} ~ (-2 \le x \lt 0)\\
        -1/4 x + 1/2 & \mathrm{if} ~ (0 \le x \lt 2)\\
        0 & \mathrm{otherwise}
    \end{array}
    \right.\\[10pt]
    h(x) &= 
    \left\{
    \begin{array}{ll}
        1 / 6 & \mathrm{if} ~ (-3 \le x \lt 3)\\
        0 & \mathrm{otherwise}
    \end{array}
    \right.
\end{align*}
$$

関数の定義

In [74]:
use std::f64::consts::PI;

fn f(x: f64) -> f64 {
    1.0 / (2.0 * PI).sqrt() * f64::exp(-x.powi(2) / 2.0)
}

fn g(x: f64) -> f64 {
    (0.5 - 0.25 * x.abs()).max(0.0)
}

fn h(x: f64) -> f64 {
    if -3.0 <= x && x < 3.0 {
        1.0 / 6.0
    } else {
        0.0
    }
}

プロット

In [92]:
const NUM: usize = 200;

// x
let x = linspace(-5.0, 5.0, NUM);

// f(x)
let fx = x.iter().copied().map(f).collect::<Vec<_>>();

// g(x)
let gx = x.iter().copied().map(g).collect::<Vec<_>>();

// h(x)
let hx = x.iter().copied().map(h).collect::<Vec<_>>();

In [93]:
evcxr_figure((600, 480), |root| {
    root.fill(&WHITE)?;

    let mut chart = ChartBuilder::on(&root)
        .caption("functions", ("Sans", 20).into_font())
        .x_label_area_size(40)
        .y_label_area_size(40)
        .build_cartesian_2d(-4.0..4.0, 0.0..0.6)?;

    chart.configure_mesh()
        .draw()?;

    chart.draw_series(
        LineSeries::new(
            x.iter().copied().zip(fx.iter().copied()),
            &BLUE
        )
    )?
    .label("y = f(x)")
    .legend(|(x, y)| Circle::new((x, y), 3, BLUE.filled()));

    chart.draw_series(
        LineSeries::new(
            x.iter().copied().zip(gx.iter().copied()),
            &RED
        )
    )?
    .label("y = g(x)")
    .legend(|(x, y)| Circle::new((x, y), 3, RED.filled()));

    chart.draw_series(
        LineSeries::new(
            x.iter().copied().zip(hx.iter().copied()),
            &GREEN
        )
    )?
    .label("y = h(x)")
    .legend(|(x, y)| Circle::new((x, y), 3, GREEN.filled()));

    // 凡例の設定
    chart.configure_series_labels()
        .position(SeriesLabelPosition::UpperRight)
        .margin(10)
        .legend_area_size(10)
        .border_style(BLACK)
        .background_style(BLACK.mix(0.1))
        .label_font(("roman", 20))
        .draw()?;

    Ok(())
})

functions
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
0.0
 
 
 
0.1
 
 
 
0.2
 
 
 
0.3
 
 
 
0.4
 
 
 
0.5
 
 
 
0.6
 
 
 
 
-4.0
 
 
 
-3.0
 
 
 
-2.0
 
 
 
-1.0
 
 
 
0.0
 
 
 
1.0
 
 
 
2.0
 
 
 
3.0
 
 
 
4.0
 
 
<polyline fill="none" opacity="1" stroke="#0000FF" stroke-width="1" points="40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 43,439 47,439 50,439 54,439 57,439 61,439 64,439 68,439 71,439 75,439 78,439 82,439 85,438 89,438 93,438 96,438 100,438 103,437 107,437 110,436 114,436 117,435 121,435 124,434 128,433 131,432 135,431 138,430 142,429 145,427 149,425 152,424 156,422 159,419 163,417 166,414 170,411 173,408 177,405 180,401 184,397 187,393 191,388 194,384 198,378 201,373 205,367 208,361 212,355 215,348 219,341 222,334 226,326 229,319 233,311 236,303 240,295 244,286 247,278 251,270 254,261 258,253 261,245 265,237 268,229 272,221 275,214 279,207 282,201 286,194 289,189 293,184 296,179 300,175 303,172 307,169 310,167 314,166 317,165 321,165 324,166 328,167 331,169 335,172 338,175 342,179 345,184 349,189 352,194 356,201 359,207 363,214 366,221 370,229 373,237 377,245 380,253 384,261 387,270 391,278 394,286 398,295 402,303 405,311 409,319 412,326 416,334 419,341 423,348 426,355 430,361 433,367 437,373 440,378 444,384 447,388 451,393 454,397 458,401 461,405 465,408 468,411 472,414 475,417 479,419 482,422 486,424 489,425 493,427 496,429 500,430 503,431 507,432 510,433 514,434 517,435 521,435 524,436 528,436 531,437 535,437 538,438 542,438 545,438 549,438 553,438 556,439 560,439 563,439 567,439 570,439 574,439 577,439 581,439 584,439 588,439 591,439 595,439 598,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 "/>
<polyline fill="none" opacity="1" stroke="#FF0000" stroke-width="1" points="40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 43,439 47,439 50,439 54,439 57,439 61,439 64,439 68,439 71,439 75,439 78,439 82,439 85,439 89,439 93,439 96,439 100,439 103,439 107,439 110,439 114,439 117,439 121,439 124,439 128,439 131,439 135,439 138,439 142,439 145,439 149,439 152,439 156,439 159,439 163,439 166,439 170,439 173,439 177,439 180,437 184,428 187,420 191,411 194,402 198,394 201,385 205,376 208,368 212,359 215,350 219,342 222,333 226,324 229,316 233,307 236,299 240,290 244,281 247,273 251,264 254,255 258,247 261,238 265,229 268,221 272,212 275,203 279,195 282,186 286,177 289,169 293,160 296,152 300,143 303,134 307,126 310,117 314,108 317,100 321,100 324,108 328,117 331,126 335,134 338,143 342,152 345,160 349,169 352,177 356,186 359,195 363,203 366,212 370,221 373,229 377,238 380,247 384,255 387,264 391,273 394,281 398,290 402,299 405,307 409,316 412,324 416,333 419,342 423,350 426,359 430,368 433,376 437,385 440,394 444,402 447,411 451,420 454,428 458,437 461,439 465,439 468,439 472,439 475,439 479,439 482,439 486,439 489,439 493,439 496,439 500,439 503,439 507,439 510,439 514,439 517,439 521,439 524,439 528,439 531,439 535,439 538,439 542,439 545,439 549,439 553,439 556,439 560,439 563,439 567,439 570,439 574,439 577,439 581,439 584,439 588,439 591,439 595,439 598,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 600,439 "/>
<polyline fill="none" opacity="1" stroke="#00FF00" stroke-width="1" points="40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,439 40,4

#### KLダイバージェンスの計算

離散値に直して積分する

In [101]:
const EPS: f64 = 1e-10;

fn kl_divergence(p: &Vec<f64>, q: &Vec<f64>) -> f64 {
    p
        .iter()
        .zip(q.iter())
        .map(|(&pi, &qi)| (
            pi / NUM as f64 + EPS,
            qi / NUM as f64 + EPS
        ))
        .map(|(pi, qi)| pi * (pi / qi).ln())
        .sum()
}

In [102]:
let funcs = [fx.clone(), gx.clone(), hx.clone()];
let labels = ["f", "g", "h"];

In [103]:
for i in 0..3 {
    for j in 0..3 {
        let kld = kl_divergence(&funcs[i], &funcs[j]);

        println!("D_KL( {} || {} ) = {}", labels[i], labels[j], kld);
    }
};

D_KL( f || f ) = 0
D_KL( f || g ) = 0.0592454658316637
D_KL( f || h ) = 0.041165007754176974
D_KL( g || f ) = 0.00586053773246323
D_KL( g || g ) = 0
D_KL( g || h ) = 0.05953179630295036
D_KL( h || f ) = 0.06421860035958628
D_KL( h || g ) = 0.5258095061012997
D_KL( h || h ) = 0


KLダイバージェンスの性質が確認できる．

- 自分自身に対してゼロを返す
  - $D_\mathrm{KL}(f || f) = 0$
  - $D_\mathrm{KL}(g || g) = 0$
  - $D_\mathrm{KL}(h || h) = 0$
- 近い分布ほど小さい値をとる
  - $D_\mathrm{KL}(g || f) = 0.00586053773246323$
- 遠い分布ほど大きい値をとる
  - $D_\mathrm{KL}(h || g) = 0.5258095061012997$